src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs

```diff
@@ -0,0 +1,129 @@
+            this.traceHelper = new EndToEndTraceHelper(loggerFactory?.CreateLogger(LoggerName) ?? throw new ArgumentNullException(nameof(loggerFactory)), false);
+        }
+
+        internal event Action<TokenRenewalState> Renewing;
+
+        internal event Action<NewTokenAndFrequency> Renewed;
+
+        internal event Action<NewTokenAndFrequency, Exception> RenewalFailed;
+
+        /// <inheritdoc />
+        public AppAuthTokenCredential Create(IConfiguration configuration, CancellationToken cancellationToken = default) =>
+            this.Create(configuration, TimeSpan.FromMinutes(5), TimeSpan.FromSeconds(30), cancellationToken);
+
+        /// <inheritdoc />
+        public AppAuthTokenCredential Create(IConfiguration configuration, TimeSpan tokenRefreshOffset, TimeSpan tokenRefreshRetryDelay, CancellationToken cancellationToken = default)
+        {
+            var context = new TokenRequestContext(new[] { "https://storage.azure.com/.default" }, null);
+
+            AzureTokenCredential tokenCredential = this.componentFactory.CreateTokenCredential(configuration);
+            AccessToken value = tokenCredential.GetToken(context, default);
+            var state = new TokenRenewalState
+            {
+                Context = context,
+                Credential = tokenCredential,
+                Previous = value,
+                RefreshDelay = tokenRefreshRetryDelay,
+                RefreshOffset = tokenRefreshOffset,
+            };
+
+            return new AppAuthTokenCredential(
+                value.Token,
+                (o, t) => this.RenewTokenAsync(o as TokenRenewalState, t),
+                state,
+                GetRenewalFrequency(value, tokenRefreshOffset));
+        }
+
+        private async Task<NewTokenAndFrequency> RenewTokenAsync(TokenRenewalState state, CancellationToken cancellationToken)
+        {
+            this.OnRenewing(state);
+
+            NewTokenAndFrequency next;
+            AccessToken result;
+            try
+            {
+                result = await state.Credential.GetTokenAsync(state.Context, cancellationToken);
+            }
+            catch (Exception e)
+            {
+                next = new NewTokenAndFrequency(state.Previous.Token, state.RefreshDelay);
+                this.OnRenewalFailed(next, e);
+                return next;
+            }
+
+            // Save the token in case renewal results in an exception
+            state.Previous = result;
+            next = new NewTokenAndFrequency(result.Token, GetRenewalFrequency(result, state.RefreshOffset));
+            this.OnRenewed(next);
+            return next;
+        }
+
+        private void OnRenewing(TokenRenewalState state) =>
+            this.Renewing?.Invoke(state);
+
+        private void OnRenewed(NewTokenAndFrequency next) =>
+            this.Renewed?.Invoke(next);
+
+        private void OnRenewalFailed(NewTokenAndFrequency next, Exception exception)
```

Comment:
I think the only problem comes down to how the error surfaces when we eventually realize we're not going to be able to get the token renewed. The token is being used by the Azure Storage client in a different thread from the thread that is renewing the token (because renewal happens via a `Timer`). So, currently we will continue to fetch the token until some operation attempts to use a stale token (now that is has expired), and throws an exception about being unable to connect to the storage account.

As of now, we'll continue to retry the connection in the background, but is it better to simply give up after some number of retries even if the worker is running?

Reply:
This behavior of continuing to retry in the background may be okay since it is on the worker and using the customer's resource. In shared infrastructure, we stop retrying 5 attempts I think. Do you know if the timer is stopped/disposed if some operation attempts to use a stale token?

Reply:
@cgillum - How do you feel about the current logic? Currently in this PR the `TokenCredential` will renew its token in the background 5 minutes before expiration (which is typically an hour). However, if it runs into a problem, it will retry after 30 seconds. If the `TaskHubWorker` attempts to interact with Azure Storage while the token is still stale, presumably it will throw an exception. I am unsure if this ultimately crashes the application, but if the app stays alive the timer will continue to retry token renewal (which presumably worked at some point lol).

Reply:
The potential problem I'd be concerned about is a resource leak if we spin up an infinite number of background refresh jobs. My understanding of the code, however, is that we'll only ever create a single `TokenCredential` object and try to reuse or refresh it indefinitely. That means we'll only ever have a single refresh loop running and no possibility of a resource leak. So no concerns there if I'm reading the code right.

If the refresh fails and the token goes stale, then we'll get a flood of Azure Storage exceptions until the app instance is recycled. It's not ideal, but it seems like there's at least a _possibility_ of recovery if the refresh operation eventually succeeds, even if it takes an hour or two. If we stop the loop, it seems to me that the app will be permanently stuck until the app recycles since we'll never create a new token. From that perspective, infinite retry feels like a better way to go.

Assuming that the refresh loop is no more likely to fail than the initial token acquisition, then I'd say I'm fine with the current design given the above assertions.

Reply:
That's right. There is one `TokenCredential` for the lifetime of the `TaskHubWorker` (although we create a new `TokenCredential` for each call to `TryGetScaleMonitor`). And I think you're right about the implications for stopping the token refresh: if we stop it, then we'll have to wait for the application to be restarted.

We _could_ have some more elaborate back-off logic in the case the current time is beyond the expiration date, and we're still attempting to renew the token (like an exponential back-off). But this may be overkill right now.